### PR TITLE
Linear Search support for Window groups queries

### PR DESCRIPTION
**Pull request by mustafasrepo**
# Which issue does this PR close?
Closes [5213](https://github.com/apache/arrow-datafusion/issues/5213).
Closes [5275](https://github.com/apache/arrow-datafusion/issues/5275).

# Rationale for this change

This is a continuation of [#4989](https://github.com/apache/arrow-datafusion/pull/4989), it extends amortized linear search support to window frames in GROUPS mode. As discussed in the preceding PR, this results in significant performance gains, which is the main purpose of this PR.

# What changes are included in this PR?
This PR removes the old GROUPS mechanism and replaces it with the amortized linear search. This reduces code size and yields performance gains simultaneously.

# Are there any user-facing changes?

No.

### CODE REVIEW

File: datafusion/physical-expr/src/window/window_frame_state.rs [hunk 1]
```diff
@@ -36,11 +34,13 @@ use std::sync::Arc;
 pub enum WindowFrameContext<'a> {
     // ROWS-frames are inherently stateless:
     Rows(&'a Arc<WindowFrame>),
+    // TODO: Fix this comment, briefly mention what the state is.
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/window_frame_state.rs:37]: Update the comment below and remove this TODO once the state issue (explained below) is fixed.

File: datafusion/physical-expr/src/window/window_frame_state.rs [hunk 2]
```diff
@@ -36,11 +34,13 @@ use std::sync::Arc;
...
     // RANGE-frames will soon have a stateful implementation that is more efficient than a stateless one:
     Range {
         window_frame: &'a Arc<WindowFrame>,
         state: WindowFrameStateRange,
     },
+    // TODO: Fix this comment, briefly mention what the state is.
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/window_frame_state.rs:43]: Update the comment below and remove this TODO once the state issue (explained below) is fixed.

File: datafusion/physical-expr/src/window/window_frame_state.rs [hunk 3]
```diff
@@ -159,6 +161,8 @@ impl<'a> WindowFrameContext<'a> {
     }
 }
 
+// TODO: Fix this struct and the comment when we move "where-we-left-off"
+//       information from arguments/return values into the state.
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/window_frame_state.rs:165]: It seems we handle state in a weird way. The "where was I?" information is passed into this function with an argument, and the new "where am I?" information is passed out via the return value. This is precisely the algorithm state, isn't it? Let's collect all state information in one place (you have an empty `struct` for this already!) and fix this.

File: datafusion/physical-expr/src/window/window_frame_state.rs [hunk 4]
```diff
@@ -193,6 +197,9 @@ impl WindowFrameStateRange {
                 }
             }
             WindowFrameBound::CurrentRow => {
+                // TODO: Is this check at the right place? Is this being empty only a problem in this
+                //       match arm? Or is it even a problem here? Let's add a test to exercise this
+                //       if it doesn't exist, and if necessary, move this check to its right place.
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/window_frame_state.rs:202]: TODO is self-explanatory, let's understand what is going on here and fix.

File: datafusion/physical-expr/src/window/window_frame_state.rs [hunk 5]
```diff
@@ -227,6 +234,9 @@ impl WindowFrameStateRange {
                     length,
                 )?,
             WindowFrameBound::CurrentRow => {
+                // TODO: Is this check at the right place? Is this being empty only a problem in this
+                //       match arm? Or is it even a problem here? Let's add a test to exercise this
+                //       if it doesn't exist, and if necessary, move this check to its right place.
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/window_frame_state.rs:239]: TODO is self-explanatory, let's understand what is going on here and fix.

File: datafusion/physical-expr/src/window/window_frame_state.rs [hunk 6]
```diff
@@ -340,679 +354,331 @@ impl WindowFrameStateRange {
...
-    /// This function calculates beginning/ending indices for the frame of the current row.
     fn calculate_range(
         &mut self,
         window_frame: &Arc<WindowFrame>,
         range_columns: &[ArrayRef],
         length: usize,
         idx: usize,
     ) -> Result<Range<usize>> {
+        // TODO: This check contradicts with the same check in the handling of CurrentRow
+        //       below. See my comment there, and fix this in the context of that.
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/window_frame_state.rs:374]: TODO is self-explanatory, let's understand what is going on here and fix.

File: datafusion/physical-expr/src/window/window_frame_state.rs [hunk 7]
```diff
@@ -340,679 +354,331 @@ impl WindowFrameStateRange {
...
-        let end = match window_frame.end_bound {
-            // UNBOUNDED PRECEDING
-            WindowFrameBound::Preceding(ScalarValue::UInt64(None)) => {
-                return Err(DataFusionError::Internal(format!(
-                    "Frame end cannot be UNBOUNDED PRECEDING '{window_frame:?}'"
-                )))
+            WindowFrameBound::CurrentRow => {
+                // TODO: Is this check at the right place? Is this being empty only a problem in this
+                //       match arm? Or is it even a problem here? Let's add a test to exercise this
+                //       if it doesn't exist, and if necessary, move this check to its right place.
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/window_frame_state.rs:397]: TODO is self-explanatory, let's understand what is going on here and fix.

File: datafusion/physical-expr/src/window/window_frame_state.rs [hunk 8]
```diff
@@ -340,679 +354,331 @@ impl WindowFrameStateRange {
...
-            WindowFrameBound::Following(ScalarValue::UInt64(None)) => length,
-            // ERRONEOUS FRAMES
-            WindowFrameBound::Preceding(_) | WindowFrameBound::Following(_) => {
-                return Err(DataFusionError::Internal(
-                    "Groups should be Uint".to_string(),
-                ))
+            WindowFrameBound::CurrentRow => {
+                // TODO: Is this check at the right place? Is this being empty only a problem in this
+                //       match arm? Or is it even a problem here? Let's add a test to exercise this
+                //       if it doesn't exist, and if necessary, move this check to its right place.
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/window_frame_state.rs:428]: TODO is self-explanatory, let's understand what is going on here and fix.

File: datafusion/physical-expr/src/window/window_frame_state.rs [hunk 9]
```diff
@@ -340,679 +354,331 @@ impl WindowFrameStateRange {
...
-        while !self.reached_end
-            && self.window_frame_end_idx <= current_window_frame_end_idx
+
+        // Update the group index `idx` belongs to:
+        while self.current_group_idx < self.group_start_indices.len()
+            && idx >= self.group_start_indices[self.current_group_idx].1
         {
-            self.advance_one_group::<SEARCH_SIDE>(range_columns)?;
+            self.current_group_idx += 1;
         }
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/window_frame_state.rs:507]: Let's make sure my simplification is not breaking anything in a corner case.

File: datafusion/physical-expr/src/window/window_frame_state.rs [hunk 10]
```diff
@@ -340,679 +354,331 @@ impl WindowFrameStateRange {
...
-                low = high;
-                step_size *= 2;
-                high += step_size;
-            } else {
-                break;
+        // Calculate index of the group boundary:
+        Ok(match (SIDE, SEARCH_SIDE) {
+            // TODO: Is it normal that window frame start and end are asymmetric? You have
+            //       PRECEDING and FOLLOWING cases separate for end, but not for start.
+            //       Seems like this is OK, but let's make sure.
```
Comment (ozankabak) [inline datafusion/physical-expr/src/window/window_frame_state.rs:539]: Your implementation seems correct to me, but such asymmetries are often indicative of subtle bugs. So, let's make sure there is no catch here. If this is safe, you can remove this TODO and explain your reasoning to me in our next meeting.

**Review by ozankabak — COMMENTED** [review]
You switched to an implementation that used `sort_options` even in GROUPS mode. I remember us discussing before that this was unnecessary and a simple equality check is enough in that context. I tried that with the new implementation and all tests seem to pass. Please verify this again and maybe add some tests for it if necessary.

If you believe our previous conclusion was wrong, we can discuss in detail in our next meeting and revert. Once you address the TODO's I left, this will be in good shape.

File: datafusion/physical-expr/src/window/window_frame_state.rs (hunk 4 above)
Comment (mustafasrepo) [reply datafusion/physical-expr/src/window/window_frame_state.rs]: I have added a comment explaining their rationale.

File: datafusion/physical-expr/src/window/window_frame_state.rs (hunk 5 above)
Comment (mustafasrepo) [reply datafusion/physical-expr/src/window/window_frame_state.rs]: I have added a comment explaining their rationale.

File: datafusion/physical-expr/src/window/window_frame_state.rs (hunk 7 above)
Comment (mustafasrepo) [reply datafusion/physical-expr/src/window/window_frame_state.rs]: This is indeed wrong. Removed it

File: datafusion/physical-expr/src/window/window_frame_state.rs (hunk 8 above)
Comment (mustafasrepo) [reply datafusion/physical-expr/src/window/window_frame_state.rs]: Removed it

**Review by ozankabak — APPROVED** [review]
Looks good, let's send upstream 👍  We will optimize the closure in a follow-on PR. Please add it to our task queue.

**Comment by mustafasrepo** [comment]
Merged Upstream

**Pull request closed without merging**